File: src/axhub-helpers/keychain.ts
/**
 * keychain.ts — token discovery from OS keychain.
 *
 * Reads the axhub access_token that ax-hub-cli stores under service="axhub"
 * via zalando/keyring. macOS: `security find-generic-password -s axhub -w`;
 * Linux: `secret-tool lookup service axhub`; Windows: PowerShell + Add-Type
 * PInvoke against advapi32!CredReadW (see ./keychain-windows.ts).
 *
 * Storage format: `go-keyring-base64:<base64 JSON>` where decoded JSON has
 * `{schema_version, access_token, token_type, expires_at, scopes}`.
 *
 * Pure functions live here so tests can import without booting the helper
 * binary's main dispatch (importing index.ts triggers process.argv parsing).
 */

import { readWindowsKeychain } from "./keychain-windows.ts";

export const parseKeyringValue = (raw: string): string | null => {
  if (raw.length === 0) return null;
  const stripped = raw.startsWith("go-keyring-base64:") ? raw.slice("go-keyring-base64:".length) : raw;
  let decoded: string;
  try {
    decoded = Buffer.from(stripped.trim(), "base64").toString("utf8");
  } catch {
    return null;
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(decoded);
  } catch {
    return null;
  }
  if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) return null;
  const obj = parsed as Record<string, unknown>;
  const tok = obj["access_token"];
  if (typeof tok !== "string" || tok.length < 16) return null;
  return tok;
};

export interface KeychainResult {
  token?: string;
  source?: string;
  error?: string;
}

export const readKeychainToken = (): KeychainResult => {
  const platform = process.platform;
  if (platform === "darwin") {
    try {
      const result = Bun.spawnSync({
        cmd: ["security", "find-generic-password", "-s", "axhub", "-w"],
        stdout: "pipe",
        stderr: "pipe",
        timeout: 5000,
      });
      if (result.exitCode !== 0) {
        return {
          error:
            "잠깐만요.\n" +
            "원인: macOS keychain 에 axhub token 이 저장돼 있지 않아요.\n" +
            "해결: 'axhub auth login' 으로 한 번 로그인해주세요.\n" +
            "다음: 로그인 후 token-init 자동 실행됩니다.",
        };
      }
      const raw = (result.stdout?.toString() ?? "").trim();
      const token = parseKeyringValue(raw);
      if (token === null)
        return {
          error:
            "이상해요.\n" +
            "원인: macOS keychain 의 axhub token 형식을 파싱할 수 없어요 (axhub CLI 버전 mismatch 가능).\n" +
            "해결: 'axhub auth login --force' 로 재발급 시도해주세요.\n" +
            "다음: 그래도 안 되면 'axhub --version' 으로 CLI 버전 확인 후 업그레이드.",
        };
      return { token, source: "macos-keychain" };
    } catch {
      return {
        error:
          "잠깐만요.\n" +
          "원인: macOS 'security' 명령 실행 자체가 실패했어요.\n" +
          "해결: /usr/bin/security 가 PATH 에 있는지 확인해주세요.\n" +
          "다음: 또는 AXHUB_TOKEN 환경변수로 우회 → export AXHUB_TOKEN=axhub_pat_...",
      };
    }
  }
  if (platform === "linux") {
    try {
      const result = Bun.spawnSync({
        cmd: ["secret-tool", "lookup", "service", "axhub"],
        stdout: "pipe",
        stderr: "pipe",
        timeout: 5000,
      });
      if (result.exitCode !== 0) {
        return {
          error:
            "잠깐만요.\n" +
            "원인: Linux secret-service 에 axhub token 이 저장돼 있지 않거나 secret-tool 미설치.\n" +
            "해결: 'sudo apt-get install libsecret-tools' 후 'axhub auth login' 실행.\n" +
            "다음: 또는 AXHUB_TOKEN 환경변수로 우회 → export AXHUB_TOKEN=axhub_pat_...",
        };
      }
      const raw = (result.stdout?.toString() ?? "").trim();
      const token = parseKeyringValue(raw);
      if (token === null)
        return {
          error:
            "이상해요.\n" +
            "원인: Linux secret-service 의 axhub token 형식을 파싱할 수 없어요 (axhub CLI 버전 mismatch 가능).\n" +
            "해결: 'axhub auth login --force' 로 재발급 시도해주세요.\n" +
            "다음: 그래도 안 되면 'axhub --version' 으로 CLI 버전 확인 후 업그레이드.",
        };
      return { token, source: "linux-secret-service" };
    } catch {
      return {
        error:
          "잠깐만요.\n" +
          "원인: secret-tool 명령이 PATH 에 없거나 D-Bus session bus 미실행.\n" +
          "해결: 'sudo apt-get install libsecret-tools' + 'eval $(dbus-launch --sh-syntax)' 시도.\n" +
          "다음: 또는 AXHUB_TOKEN 환경변수로 우회 → export AXHUB_TOKEN=axhub_pat_...",
      };
    }
  }
  if (platform === "win32") return readWindowsKeychain();
  return {
    error:
      "잠깐만요.\n" +
      `원인: 지원하지 않는 플랫폼이에요 (platform=${platform}).\n` +
      "해결: AXHUB_TOKEN 환경변수로 우회 가능해요.\n" +
      "다음: export AXHUB_TOKEN=axhub_pat_... 후 token-init 재시도.",
  };
};

File: crates/axhub-helpers/src/keychain.rs
use base64::Engine;
use serde::{Deserialize, Serialize};

use crate::keychain_windows::read_windows_keychain;

pub fn parse_keyring_value(raw: &str) -> Option<String> {
    if raw.is_empty() {
        return None;
    }
    let stripped = raw.strip_prefix("go-keyring-base64:").unwrap_or(raw).trim();
    let decoded = base64::engine::general_purpose::STANDARD
        .decode(stripped)
        .ok()?;
    let parsed: serde_json::Value = serde_json::from_slice(&decoded).ok()?;
    let tok = parsed.get("access_token")?.as_str()?;
    (tok.len() >= 16).then(|| tok.to_string())
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct KeychainResult {
    pub token: Option<String>,
    pub source: Option<String>,
    pub error: Option<String>,
}
impl KeychainResult {
    pub fn token(token: impl Into<String>, source: impl Into<String>) -> Self {
        Self {
            token: Some(token.into()),
            source: Some(source.into()),
            error: None,
        }
    }
    pub fn error(error: impl Into<String>) -> Self {
        Self {
            token: None,
            source: None,
            error: Some(error.into()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CommandOutput {
    pub exit_code: i32,
    pub stdout: String,
    pub stderr: String,
}
pub type CommandRunner = fn(&[&str], u64) -> CommandOutput;

pub fn default_runner(cmd: &[&str], _timeout_ms: u64) -> CommandOutput {
    match crate::spawn::spawn_sync(cmd) {
        Ok(r) => CommandOutput {
            exit_code: r.exit_code.unwrap_or(1),
            stdout: r.stdout,
            stderr: r.stderr,
        },
        Err(e) => CommandOutput {
            exit_code: 1,
            stdout: String::new(),
            stderr: e.to_string(),
        },
    }
}

pub fn read_keychain_token() -> KeychainResult {
    read_keychain_token_with_runner(std::env::consts::OS, default_runner)
}

pub fn read_keychain_token_with_runner(platform: &str, runner: CommandRunner) -> KeychainResult {
    match platform {
        "macos" => read_unix_keychain(&["security", "find-generic-password", "-s", "axhub", "-w"], 5000, "macos-keychain", runner,
            "잠깐만요.\n원인: macOS keychain 에 axhub token 이 저장돼 있지 않아요.\n해결: 'axhub auth login' 으로 한 번 로그인해주세요.\n다음: 로그인 후 token-init 자동 실행됩니다."),
        "linux" => read_unix_keychain(&["secret-tool", "lookup", "service", "axhub"], 5000, "linux-secret-service", runner,
            "잠깐만요.\n원인: Linux secret-service 에 axhub token 이 저장돼 있지 않거나 secret-tool 미설치.\n해결: 'sudo apt-get install libsecret-tools' 후 'axhub auth login' 실행.\n다음: 또는 AXHUB_TOKEN 환경변수로 우회 → export AXHUB_TOKEN=axhub_pat_..."),
        "windows" => read_windows_keychain(),
        other => KeychainResult::error(format!("잠깐만요.\n원인: 지원하지 않는 플랫폼이에요 (platform={other}).\n해결: AXHUB_TOKEN 환경변수로 우회 가능해요.\n다음: export AXHUB_TOKEN=axhub_pat_... 후 token-init 재시도.")),
    }
}

fn read_unix_keychain(
    cmd: &[&str],
    timeout_ms: u64,
    source: &str,
    runner: CommandRunner,
    not_found: &str,
) -> KeychainResult {
    let result = runner(cmd, timeout_ms);
    if result.exit_code != 0 {
        return KeychainResult::error(not_found);
    }
    match parse_keyring_value(result.stdout.trim()) {
        Some(token) => KeychainResult::token(token, source),
        None => KeychainResult::error("이상해요.\n원인: keychain 의 axhub token 형식을 파싱할 수 없어요 (axhub CLI 버전 mismatch 가능).\n해결: 'axhub auth login --force' 로 재발급 시도해주세요.\n다음: 그래도 안 되면 'axhub --version' 으로 CLI 버전 확인 후 업그레이드."),
    }
}
